src/workerd/api/node/zlib-util.c++

```diff
@@ -878,6 +837,126 @@ void ZlibUtil::CompressionStream<CompressionContext>::FreeForZlib(void* data, vo
+      opts.windowBits.orDefault(Z_DEFAULT_WINDOWBITS), opts.memLevel.orDefault(Z_DEFAULT_MEMLEVEL),
+      opts.strategy.orDefault(Z_DEFAULT_STRATEGY), kj::mv(opts.dictionary));
+
+  auto flush = opts.flush.orDefault(Z_NO_FLUSH);
+  JSG_REQUIRE(Z_NO_FLUSH <= flush && flush <= Z_TREES, RangeError, "invalid flush value");
+
+  auto finishFlush = opts.finishFlush.orDefault(Z_FINISH);
+  JSG_REQUIRE(
+      Z_NO_FLUSH <= finishFlush && finishFlush <= Z_TREES, RangeError, "invalid finishFlush value");
+
+  ctx.setFlush(finishFlush);
+  ctx.setInputBuffer(getInputFromSource(data));
+  return syncProcessBuffer(ctx, result);
+}
+
+void ZlibUtil::zlibWithCallback(
+    jsg::Lock& js, InputSource data, Options options, ZlibModeValue mode, CompressCallback cb) {
+  try {
+    cb(js, kj::none, zlibSync(kj::mv(data), kj::mv(options), mode));
+  } catch (kj::Exception& ex) {
+    auto tunneledError = jsg::tunneledErrorType(ex.getDescription());
+    cb(js, tunneledError.message, kj::none);
+  }
+}
+
+template <typename Context>
+kj::Array<kj::byte> ZlibUtil::brotliSync(InputSource data, BrotliOptions opts) {
+  ContextRAII<Context> ctx(Context::Mode);
+
+  auto chunkSize = opts.chunkSize.orDefault(ZLIB_PERFORMANT_CHUNK_SIZE);
+  auto maxOutputLength = opts.maxOutputLength.orDefault(Z_MAX_CHUNK);
+
+  JSG_REQUIRE(Z_MIN_CHUNK <= chunkSize && chunkSize <= Z_MAX_CHUNK, Error, "Invalid chunkSize");
+  JSG_REQUIRE(maxOutputLength <= Z_MAX_CHUNK, Error, "Invalid maxOutputLength");
+  GrowableBuffer result(ZLIB_PERFORMANT_CHUNK_SIZE, maxOutputLength);
+
+  // TODO(soon): should we track them brotli allocationz?
+  auto maybeError = ctx.initialize(nullptr, nullptr, nullptr);
+
+  KJ_IF_SOME(err, maybeError) {
+    JSG_FAIL_REQUIRE(Error, err.message);
+  }
+
+  KJ_IF_SOME(params, opts.params) {
+    for (const auto& field: params.fields) {
+      maybeError = ctx.setParams(field.name.parseAs<int>(), field.value);
+      KJ_IF_SOME(err, maybeError) {
+        JSG_FAIL_REQUIRE(Error, err.message);
+      }
+    }
+  }
+
+  auto flush = opts.flush.orDefault(BROTLI_OPERATION_PROCESS);
+  JSG_REQUIRE(BROTLI_OPERATION_PROCESS <= flush && flush <= BROTLI_OPERATION_EMIT_METADATA,
+      RangeError, "invalid flush value");
+
+  auto finishFlush = opts.finishFlush.orDefault(BROTLI_OPERATION_FINISH);
+  JSG_REQUIRE(
+      BROTLI_OPERATION_PROCESS <= finishFlush && finishFlush <= BROTLI_OPERATION_EMIT_METADATA,
+      RangeError, "invalid finishFlush value");
+
+  ctx.setFlush(finishFlush);
+  ctx.setInputBuffer(getInputFromSource(data));
+  return syncProcessBuffer(ctx, result);
+}
+
+template <typename Context>
+void ZlibUtil::brotliWithCallback(
+    jsg::Lock& js, InputSource data, BrotliOptions options, CompressCallback cb) {
+  try {
+    cb(js, kj::none, brotliSync<Context>(kj::mv(data), kj::mv(options)));
+  } catch (kj::Exception& ex) {
+    auto tunneledError = jsg::tunneledErrorType(ex.getDescription());
+    cb(js, tunneledError.message, kj::none);
```

Comment:
Same issue as above... The `cb(...)` should only every be called once and this should likely be using `js.tryCatch(...)` instead.